git-protocol/src/fetch.rs

```diff
@@ -0,0 +1,493 @@
+
+    /// [Extra Parameters][extra] to send with the initial transport header.
+    ///
+    /// [extra]: https://git.kernel.org/pub/scm/git/git.git/tree/Documentation/technical/pack-protocol.txt#n52
+    pub extra_params: Vec<(String, Option<String>)>,
+
+    /// Prefixes of refs to ask the server to advertise via `ls-refs`.
+    ///
+    /// If this is empty, the `ls-refs` command will not be sent.
+    pub ref_prefixes: Vec<BString>,
+
+    /// Known refs to ask the server to include in the packfile.
+    ///
+    /// If both `want_refs` and `ref_prefixes` are empty, it is unlikely that
+    /// the server will send a packfile at all.
+    pub want_refs: Vec<WantRef>,
+}
+
+/// What to do with the `packfile` response.
+///
+/// _This is mostly the same as [`Delegate`], but without incurring the
+/// [`DelegateBlocking`] super-trait constraint. We can simply make [`Fetch`]
+/// parametric over the packfile sink._
+#[async_trait(?Send)]
+pub trait ReceivePack {
+    type Output;
+
+    async fn receive_pack(
+        &self,
+        pack: impl AsyncBufRead + Unpin + 'async_trait,
+        progress: impl Progress,
+    ) -> io::Result<Self::Output>;
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct ReceivePackOptions {
+    /// How many threads the packfile indexer is allowed to spawn. `None` means
+    /// unlimited.
+    pub max_indexer_threads: Option<usize>,
+    /// The maximum size in bytes of the packfile.
+    ///
+    /// If the remote sends a larger file, the transfer will be aborted.
+    pub max_pack_bytes: u64,
+}
+
+impl Default for ReceivePackOptions {
+    fn default() -> Self {
+        Self {
+            max_indexer_threads: Some(1),
+            max_pack_bytes: u64::MAX,
+        }
+    }
+}
+
+pub type PackReceived = git_pack::bundle::write::Outcome;
+
+/// The default [`ReceivePack`].
+///
+/// Writes the packfile into the given output directory, along with a v2 index.
+/// The packfile is verified.
+pub struct DefaultReceivePack {
+    opt: ReceivePackOptions,
+    out: Option<PathBuf>,
+    stop: AtomicBool,
+}
+
+impl DefaultReceivePack {
+    pub fn new(opt: ReceivePackOptions, out: PathBuf) -> Self {
+        Self {
+            opt,
+            out: Some(out),
+            stop: AtomicBool::new(false),
+        }
+    }
+}
+
+impl Drop for DefaultReceivePack {
+    fn drop(&mut self) {
+        self.stop.store(true, Ordering::Release);
+    }
+}
+
+#[async_trait(?Send)]
+impl ReceivePack for DefaultReceivePack {
+    type Output = PackReceived;
+
+    async fn receive_pack(
+        &self,
+        pack: impl AsyncBufRead + Unpin + 'async_trait,
+        prog: impl Progress,
+    ) -> io::Result<Self::Output> {
+        use git_pack::{bundle::write::Options, data::input::Mode, index::Version, Bundle};
+
+        let opts = Options {
+            thread_limit: self.opt.max_indexer_threads,
+            index_kind: Version::V2,
+            iteration_mode: Mode::Verify,
+        };
+        Bundle::write_to_directory(
```

Comment:
⚠️Big Warning⚠️: Thin packs can't yet be resolved here, even though the machinery is in place it's not yet forwarded and/or fully implemented. The reason for that is that `gitoxide` only supports clones for now where packs are never thin.

I put that [on the list](https://github.com/Byron/gitoxide/issues/104) now to not forget.

Reply:
Based on your current priorities, do you think I should use `libgit2` for now?

Reply:
If there is a more-or-less equivalent call in `libgit2` to write a pack without much hassle, so using it is low-cost, I'd be inclined to say yes just to avoid me being a bottleneck.

Technically everything is prepared to support thin-pack resolution in _the machinery_, but it's not hooked up yet as there was no need previously. I put the `git-pack` improvements all the way up so it's done right after git-ref writing is finished.

Reply:
Tomorrow will be dedicated to thin-pack resolution, I wouldn't be surprised it's available when you begin your day. Stay tuned :).

Reply:
Thin pack resolution is nearly there and [will look like this](https://github.com/Byron/gitoxide/blob/9f43bf029624f7c94346646465e366609b89e2e1/git-pack/tests/pack/data/output/count_and_entries.rs#L367). It's implemented [as an iterator](https://github.com/Byron/gitoxide/blob/9f43bf029624f7c94346646465e366609b89e2e1/git-pack/src/data/input/lookup_ref_delta_objects.rs#L40) which injects the looked-up entries into an existing pack-entry iteration.

However, even though I took my shot I wasn't lucky enough to get it right from the start and will need to add the utterly missing unit test tomorrow to figure out what part of the offset computations are off by one ;). This also means that I expect this to be ready by tomorrow.

Reply:
So, even though it should be possible to recalculate pack-offsets and offset-deltas on the fly in the light of changing pack entry sizes due to entry-insertions and other size changes, for tonight I have to declare defeat :/.

If it would have worked, you would be the first to actually receive and write thin packs with `gitoxide`, as fetch-related tests and capabilities are still missing entirely, so in any case there is quite some work ahead to validate all of this.

Reply:
As per 0ed3c655322b4c03ceac4b9c0d148d6e9948b939, it's not _too_ painful to use libgit2 in the meantime. It does lock the odb (for whatever reason), so we need to keep an owned `Repository` around.

Reply:
It's probably not a terrible idea to have `libgit2` ready at the flick of a feature toggle either, yet I hope that this code paths can soon be removed entirely.

This thin-pack problem has been annoying me in the past days but now I finally see the implementation to come to fruition. It's amazing how much work has to be done to do this correctly. The biggest hurdle is the absolutely unnecessary number of objects in the pack header, which causes extra work for particular advantage especially since the file content is secured by a trailing hash anyway. Definitely something I would change in a V3 pack format.

Without having studied how git handles thin pack resolution at all I can at least say that the `gitoxide` implementation is fully streaming even with thin-pack resolution, so resolving referenced base objects happens on the fly as well as writing the pack to disk, all while it's being indexed in preparation for producing the hashes required for the pack index file. When cloning `gitoxide` is already faster than git, and when fetching I think it will be the same.

Reply:
With e7b9d9613874cd1ebaf740dc08db467c461a4751 thin packs can be resolved. 
The bundle writer now [has an additional parameter](https://github.com/Byron/gitoxide/blob/1852e3ea98a462958862ab05f110649e3b06e2b5/git-pack/src/bundle/write/mod.rs#L38) to allow resolving arbitrary objects using an object db. Typically [it's implemented like this](https://github.com/Byron/gitoxide/blob/990ea4866be2d22ae2043da2dcd9577b748de255/git-pack/tests/pack/data/output/count_and_entries.rs#L373-L375) if a thin pack is in the realm of possibilities.

I think it has all the tests it needs already to assure it's working as expected.
The pack-generation machinery was also update to be able to produce thin packs itself which is where most of the thorough validation is coming from right now.